operator/pkg/controllers/controller.go

```diff
@@ -41,69 +42,70 @@ type GenericController struct {
 func (c *GenericController) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
 	// 1. Read Spec
 	resource := c.For()
-	if err := c.Get(ctx, req.NamespacedName, resource); err != nil {
+	var err error
+	if err = c.Get(ctx, req.NamespacedName, resource); err != nil {
 		if errors.KubeObjNotFound(err) {
 			return reconcile.Result{}, nil
 		}
-		return reconcile.Result{}, err
+		return *status.Failed, err
 	}
 	// 2. Copy object for merge patch base
 	persisted := resource.DeepCopyObject()
-	// 3. reconcile else finalize if object is deleted
-	result, err := c.reconcile(ctx, resource)
-	if err != nil {
-		resource.StatusConditions().MarkFalse(v1alpha1.Active, "", err.Error())
-		if errors.SafeToIgnore(err) {
-			return result, nil
-		}
-		zap.S().Errorf("Failed to reconcile kind %s, %v", resource.GetObjectKind().GroupVersionKind().Kind, err)
-		return reconcile.Result{Requeue: true}, err
+	// 3. Reconcile else finalize if object is deleted
+	result, reconcileErr := c.reconcile(ctx, resource, persisted)
+	// 4. Update Status using a merge patch, we want to set status even when reconcile errored
+	if err := c.Status().Patch(ctx, resource, client.MergeFrom(persisted)); err != nil && !errors.KubeObjNotFound(err) {
+		return *status.Failed, fmt.Errorf("status patch for %s, %w,", req.NamespacedName, err)
 	}
-	resource.StatusConditions().MarkTrue(v1alpha1.Active)
-	// 4. Update Status using a merge patch
-	if err := c.Status().Patch(ctx, resource, client.MergeFrom(persisted)); err != nil {
-		return reconcile.Result{}, fmt.Errorf("failed to persist changes to %s, %w", req.NamespacedName, err)
+	if reconcileErr != nil {
+		zap.S().Errorf("Error reconciling the resource %v", err)
+		return *status.Failed, reconcileErr
 	}
 	return result, nil
 }
 
-func (c *GenericController) reconcile(ctx context.Context, resource Object) (reconcile.Result, error) {
+func (c *GenericController) reconcile(ctx context.Context, resource Object, persisted runtime.Object) (reconcile.Result, error) {
+	var result *reconcile.Result
```

Comment:
If you're naming these, consider just using named return types.

Reply:
In order to return the result, need to deference the result from the controller. So can't use the named return types.

Reply:
Oh because the controller returns a pointer. That may be a decent signal that you should just keep the recommended controller-runtime API (reconcile.Result, error) for all your controllers. We eventually came back to them. Totally optional -- do what feels best.

Reply:
I went from `(reconcile.Result, error)` to returning `(*reconcile.Result, error)` as it makes the controller code simpler (when we had large number of them). There is this quirk here that we need to deference the result.